File: c++/graphs/simple_graph.cpp
// File: simple_graph.cpp
// Author: Nick Gingerella
// Description:
// Implementation of a basic graph data structure and various graph algorithms
#include <iostream>
#include <map>
#include <set>
#include <vector>
#include <string>

// A graph is
// a set of nodes/vertices
// a set of edges (pairs of vertices)

class Graph
{
    private:

    protected:
        // a vertext in the graph contains:
        // a label: name to identify vertex
        // a degree: number of edges connected to vertex
        using vertex =  struct vertex
        {
            std::string label;
            int degree;
        };

        struct VertexCompare
        {
            bool operator() (const vertex& lhs, const vertex& rhs)
            {
                return lhs.label < rhs.label;
            }
        };
        
        // map where keys are vertices, and values are lists
        // of edges from the vertex (adjacency list).
        std::map<vertex, std::vector<vertex>, VertexCompare> m_adjacencyList;

        // type aliases
        using adjacencyListItr = std::map<vertex, std::vector<vertex>>::iterator;
        using edge = std::pair<vertex,vertex>;
        using edgeListItr = std::vector<edge>;

    public:

        Graph()
        {
            std::cout << "Graph Constructor" << std::endl;
        }

        ~Graph()
        {
            std::cout << "Graph Destructor" << std::endl;
        }

        // get vertices
        std::vector<vertex> get_vertices()
        {
            std::vector<vertex> retval;
            for (auto const& entry: m_adjacencyList) {
                retval.push_back(entry.first);
            }
            return retval;
        }

        // get edges
        std::vector<edge> get_edges()
        {
            std::vector<edge> retval;
            for (auto vertex_entry: m_adjacencyList)
            {
                std::cout << vertex_entry.first.label << std::endl;
                for(auto adjacent_vertex: vertex_entry.second)
                {
                    edge e = std::make_pair(vertex_entry.first, adjacent_vertex);
                    retval.push_back(e);
                }
            }

            return retval;
        }

        // add vertex
        bool add_vertex(std::string label)
        {
            int oldSize = m_adjacencyList.size();
            std::vector<vertex> e;
            vertex v = {label, 0};
            m_adjacencyList.insert({v,e});
            int newSize = m_adjacencyList.size();

            // if the size of the list increased, then the vertex
            // was added successfully
            if (oldSize < newSize)
            {
                return true;
            }

            return false;
        }

        // remove vertex*


        // add edge
        virtual bool add_edge(std::string label1, std::string label2)
        {
            
            adjacencyListItr itr_v1 = m_adjacencyList.find({label1});
            adjacencyListItr itr_v2 = m_adjacencyList.find({label2});
            if (itr_v1 != m_adjacencyList.end() && itr_v2 != m_adjacencyList.end())
            {
                // Both vertices exist in graph, this is a valid edge

                // first vertex in edge is a valid node, add edge to 
                // this node's list
                m_adjacencyList[{label1}].push_back({label2});
                m_adjacencyList[{label2}].push_back({label1});
            }

            return true;
        }


        // remove edge*
};

class DirectedGraph: public Graph
{
    private:

        // // type aliases
        // using adjacencyListItr = std::map<vertex, std::vector<vertex>>::iterator;
        // using edge = std::pair<vertex,vertex>;
        // using edgeListItr = std::vector<edge>;

    public:
    DirectedGraph()
    {
        std::cout << "Directed Graph Constructor" << std::endl;
    }
    
    ~DirectedGraph()
    {
        std::cout << "Directed Graph Destructor" << std::endl;
    }

    bool add_edge(std::string fromLabel, std::string toLabel) override
    {
        adjacencyListItr itr_v1 = m_adjacencyList.find({fromLabel});
        adjacencyListItr itr_v2 = m_adjacencyList.find({toLabel});
        if (itr_v1 != m_adjacencyList.end() && itr_v2 != m_adjacencyList.end())
        {
            // add edge from first node to secondn node (directed)
            m_adjacencyList[{fromLabel}].push_back({toLabel});
        }
        else
        {
            return false;
        }
        

        return true;
    }

};

int main (int argc, char** argv)
{
    Graph test_graph;
    std::cout << "test_graph instantiated" << std::endl;
    test_graph.add_vertex("A");
    test_graph.add_vertex("B");
    test_graph.add_vertex("C");
    test_graph.add_vertex("D");
    for (auto const& v : test_graph.get_vertices())
    {
        std::cout << v.label << "," << v.degree << std::endl;
    }

    test_graph.add_edge("A","B");
    test_graph.add_edge("B","C");
    test_graph.add_edge("B","D");

    std::cout << "Edges" << std::endl;
    for (auto const& e: test_graph.get_edges())
    {
        std::cout << e.first.label << e.second.label << std::endl;
    }



    DirectedGraph test_graph2;
    std::cout << "test_graph2 instantiated" << std::endl;
    test_graph2.add_vertex("A");
    test_graph2.add_vertex("B");
    test_graph2.add_vertex("C");
    test_graph2.add_vertex("D");
    for (auto const& v : test_graph2.get_vertices())
    {
        std::cout << v.label << "," << v.degree << std::endl;
    }

    test_graph2.add_edge("A","B");
    test_graph2.add_edge("B","C");
    test_graph2.add_edge("B","D");

    std::cout << "Edges" << std::endl;
    for (auto const& e: test_graph2.get_edges())
    {
        std::cout << e.first.label << e.second.label << std::endl;
    }

    return 0;
}

File: python/graphs/simple_graph.py
# File: simple_graph.py
# Author: Nick Gingerella
# Description:
# Simple implementation of a graph data structure and various graph algorithms
from collections import deque

# adjacency list
graph = {
    'A': ['B', 'C'], # node A connects to nodes B and C
    'B': ['D', 'E'], # node B connects to nodes D and E
    'C': ['B'],
    'D': ['E']
}

def find_path(graph, start, end, path=[]):
    path = path + [start] # creates a new list with start node to path
    if start == end:
        return path # start and end refer to the same node
    if start not in graph:
        return None # the start node does not exist on the graph
    
    for node in graph[start]:
        if node not in path:
            newpath  = find_path(graph, node, end, path)
            if newpath:
                return newpath
    return None


def find_all_paths(graph, start, end, path=[]):
    path = path + [start] # creates a new list with start node to path
    if start == end:
        return [path] # start and end refer to the same node
    if start not in graph:
        return []
    paths = []
    for node in graph[start]:
        if node not in path:
            newpaths  = find_all_paths(graph, node, end, path)
            for newpath in newpaths:
                paths.append(newpath)
    return paths

""" Python Class
A simple Python graph class, demonstrating the essential
facts and functionalities of graphs
"""

class Graph():
    def __init__(self, graph_dict=None):
        """ init graph object
            If no dictionary or None is given,
            an empty dictionary will be used
        """
        if graph_dict == None:
            graph_dict = {}
        self.__graph_dict = graph_dict

    
    def vertices(self):
        """ returns the vertices of the graph """
        return list(self.__graph_dict.keys())
    
    def edges(self):
        """ returns the edges of the graph """
        return list(self.__generate_edges())
    
    def add_vertex(self, vertex):
        """ If the vertex "vertex" is not in
            self.__graph_dict, a key "vertex" with an empty
            list as a value is added to the dictionary.
            Otherwise nothing has to be done
        """
        if vertex not in self.__graph_dict:
            self.__graph_dict[vertex] = []
    
    def add_edge(self, edge):
        """ assumes that edge is a type set, tuple or list;
            between two vertices can be multiple edges.
        """
        edge = set(edge)
        (vertex1, vertex2) = tuple(edge)
        if vertex1 in self.__graph_dict:
            self.__graph_dict[vertex1].append(vertex2)
        else:
            self.__graph_dict[vertex1] = [vertex2]

    def get_adj_nodes(self, node):
        return self.__graph_dict[node]

    def __generate_edges(self):
        """ A static method for generating the edges of the
            graph "graph". Edges are represented as sets with
            one (a loop back to the vertex) or two vertices
        """
        edges = []
        for vertex in self.__graph_dict:
            for neighbor in self.__graph_dict[vertex]:
                if (neighbor, vertex) not in edges:
                    edges.append((vertex, neighbor))
        return edges

    def __str__(self):
        res = "vertices: "
        for k in self.__graph_dict:
            res += str(k) + " "
        res += "\nedges: "
        for edge in self.__generate_edges():
            res += str(edge) + " "
        return res

if __name__ == "__main__":
    g = {
        'a':['b','c','d'],
        'b':['a','e'],
        'c':['a','f'],
        'd':['g','a'],
        'e':['b','f'],
        'f':['c','g'],
        'g':['d','f']
    }
    graph = Graph(g)
    print("graph:")
    print(graph)
